book/src/dev/rfcs/xxxx-block-subsidy.md

```diff
@@ -0,0 +1,296 @@
+- `BLOSSOM_POW_TARGET_SPACING_RATIO: u64`
+- `PRE_BLOSSOM_HALVING_INTERVAL: Height`
+- `POST_BLOSSOM_HALVING_INTERVAL: Height`
+- `FOUNDERS_FRACTION_DIVISOR: u64`
+- `FOUNDER_ADDRESS_CHANGE_INTERVAL: u64`
+- `FOUNDER_ADDRESSES_MAINNET: [&'static str; 48]`
+- `FOUNDER_ADDRESSES_TESTNET: [&'static str; 48]`
+
+### Funding streams parameter constants
+
+The design suggests to implement the parameters needed for funding streams as:
+
+```
+/// The funding stream receiver categories
+#[allow(missing_docs)]
+pub enum FundingStreamReceiver {
+    ElectricCoinCompany,
+    ZcashFoundation,
+    MajorGrants,
+}
+
+/// The numerator for each funding stream receiving category 
+/// as described in [protocol specification §7.9.1][7.9.1].
+/// 
+/// [7.9.1]: https://zips.z.cash/protocol/protocol.pdf#zip214fundingstreams
+const FUNDING_STREAM_RECEIVER_NUMERATORS: &[(FundingStreamReceiver, u64)] = &[
+    (FundingStreamReceiver::ElectricCoinCompany, 7),
+    (FundingStreamReceiver::ZcashFoundation, 5),
+    (FundingStreamReceiver::MajorGrants, 8),
+];
+
+/// Denominator as described in [protocol specification §7.9.1][7.9.1].
+/// 
+/// [7.9.1]: https://zips.z.cash/protocol/protocol.pdf#zip214fundingstreams
+pub const FUNDING_STREAM_RECEIVER_DENOMINATOR: u64 = 100;
+
+#[allow(missing_docs)]
+pub enum FundingStreamRange {
+    StartHeight,
+    EndHeight,
+}
+
+/// Start and end Heights for funding streams 
+/// as described in [protocol specification §7.9.1][7.9.1].
+/// 
+/// [7.9.1]: https://zips.z.cash/protocol/protocol.pdf#zip214fundingstreams
+const FUNDING_STREAM_HEIGHT_RANGES: &[(Network, Range<_>)] = &[
+    (Network::Mainnet, Height(1_046_400)..Height(2_726_400)),
+    (Network::Testnet, Height(1_028_500)..Height(2_796_000)),
+];
+```
+
+Each `FundingStreamReceiver` for the mainnet must have 48 payment addresses associated with and 51 for the testnet. 
+
+https://zips.z.cash/zip-0214#mainnet-recipient-addresses
+
+The following constants are needed:
+
+- `FUNDING_STREAM_ECC_ADDRESSES_MAINNET: [&'static str; 48]`
+- `FUNDING_STREAM_ZF_ADDRESSES_MAINNET: [&'static str; 48]`
+- `FUNDING_STREAM_MG_ADDRESSES_MAINNET: [&'static str; 48]`
+- `FUNDING_STREAM_ECC_ADDRESSES_TESTNET: [&'static str; 51]`
+- `FUNDING_STREAM_ZF_ADDRESSES_TESTNET: [&'static str; 51]`
+- `FUNDING_STREAM_MG_ADDRESSES_TESTNET: [&'static str; 51]`
+
+## General subsidy
+
+The block subsidy and miner reward among other utility functions are inside the general subsidy category.
+
+https://zips.z.cash/protocol/canopy.pdf#subsidyconcepts
+
+https://zips.z.cash/protocol/canopy.pdf#subsidies
+
+- `block_subsidy(Height, Network) -> Result<Amount<NonNegative>, Error>` - Total block subsidy.
+- `miner_subsidy(Height, Network) -> Result<Amount<NonNegative>, Error>` - Miner portion.
+- `miner_fees(&Block) -> Result<Amount<NonNegative>, Error>` - Total block fees calculated as [Transparent value pool calculation](#transparent-value-pool-calculation) that belongs to the miner.
```

Comment:
For the implementation i think we are going to need to have `state` in this function somehow(not sure if passing it as function argument is the best idea). Not needed for sapling as we have `value_balance` in the same transaction but for the other 2 types we need to get an output from a previous block given an input `outpoint` in this block. I think `zebra_state::Request::AwaitUtxo()` will do it after we have the service in place.

Reply:
You need chain state (at least, the ability to look up previous transactions) to obtain input values for the transparent components of the transaction (transparent inputs, and TZE inputs if [ZIP 222](https://zips.z.cash/zip-0222) is deployed).

For the shielded components, everything you need to compute the transaction fee is part of the transaction:
- Sprout: `vpub_old` and `vpub_new` values for each JoinSplit.
- Sapling: `valueBalance`.
- A future new shielded pool would likely have its own `valueBalance` field.

Reply:
A full transaction index is a lot of work and data (in `zcashd` you have to explicitly enable it with `-txindex`), and you usually only care about determining the fee when actually mining a transaction, so looking up the UTXO state when a block is being connected to a chain is probably sufficient.

Reply:
> I think `zebra_state::Request::AwaitUtxo()` will do it after we have the service in place.

Zebra splits verification into `semantic verification` and `contextual verification` stages:
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0002-parallel-verification.md#definitions

If a check requires chain state, then it is a `contextual verification`, and it belongs in `zebra-state`.

(We can write functions that take a UTXO input in `zebra-consensus`, to keep them with the rest of the subsidy functions. But those functions can only be called from the `StateService` in `zebra-state`.)

Let's split the RFC into semantic and contextual verification, and focus on semantic verification right now?

Reply:
Contextual validation can't be implemented in `zebra-consensus`.

```suggestion
```